File: jdroid-gradle-googleplay-publisher-plugin/src/main/groovy/com/jdroid/android/googleplay/publisher/commons/AbstractTask.java
package com.jdroid.android.googleplay.publisher.commons;

import org.gradle.api.DefaultTask;
import org.gradle.api.tasks.TaskAction;

public abstract class AbstractTask extends DefaultTask {
	
	protected PropertyResolver propertyResolver;
	
	@TaskAction
	public final void doExecute() {
		propertyResolver = new PropertyResolver(getProject());
		onExecute();
	}
	
	protected abstract void onExecute();
	
}

File: google-play-api-plugin/src/main/java/com/dipien/google/play/api/commons/AbstractTask.kt
package com.dipien.google.play.api.commons

import com.dipien.google.play.api.GooglePlayPublisherPlugin
import com.dipien.google.play.api.GooglePlayPublisherPluginExtension
import org.gradle.api.DefaultTask
import org.gradle.api.tasks.Internal
import org.gradle.api.tasks.TaskAction

abstract class AbstractTask : DefaultTask() {

    @get:Internal
    protected lateinit var propertyResolver: PropertyResolver

    @TaskAction
    fun doExecute() {
        propertyResolver = PropertyResolver(project)
        onExecute()
    }

    @Internal
    protected fun getExtension(): GooglePlayPublisherPluginExtension {
        return project.extensions.getByName(GooglePlayPublisherPlugin.EXTENSION_NAME) as GooglePlayPublisherPluginExtension
    }

    protected abstract fun onExecute()
}
